src/services/ContributorAuditService/providers/Bitbucket/BitbucketApiClient.ts

```diff
@@ -0,0 +1,248 @@
+}
+
+export interface BaseBitBucketApiResponse {
+  type: (typeof BitbucketResponseTypes)[keyof typeof BitbucketResponseTypes];
+  values: BitbucketRepository[] | BitbucketCommit[];
+  next?: string;
+}
+
+export const BitbucketResponseTypes = {
+  Commit: "commit",
+  Repository: "repository",
+} as const;
+
+// NOTE: Bitbucket related interfaces do not represent the full response from Bitbucket, only the fields we care about
+export interface BitBucketRepositoryApiResponse extends BaseBitBucketApiResponse {
+  type: typeof BitbucketResponseTypes.Repository;
+  values: BitbucketRepository[];
+}
+
+export interface BitbucketRepository {
+  uuid: string;
+  full_name: string;
+  name: string;
+  is_private: boolean;
+  updated_on: string;
+  workspace: BitbucketWorkspace;
+}
+
+export interface BitbucketWorkspace {
+  type: string;
+  uuid: string;
+  name: string;
+  slug: string;
+}
+
+export interface BitbucketCommitsApiResponse extends BaseBitBucketApiResponse {
+  type: typeof BitbucketResponseTypes.Commit;
+  values: BitbucketCommit[];
+}
+
+export interface BitbucketCommit {
+  author: BitbucketAuthor;
+  date: string;
+}
+
+export interface BitbucketAuthor {
+  raw: string;
+  name: string;
+  emailAddress: string;
+  displayName: string;
+  user: BitbucketUser;
+}
+
+interface BitbucketUser {
+  display_name: string;
+  type: string;
+  nickname: string;
+}
+
+class BitbucketApiClient {
+  private readonly client: AxiosInstance;
+  private readonly workspace: string;
+  private readonly days: number;
+  private readonly dateToFilter: string;
+
+  constructor(
+    days: number,
+    username: string,
+    password: string,
+    workspace: string,
+    baseUri: string = SOOS_BITBUCKET_CONTRIBUTOR_AUDIT_CONSTANTS.Urls.API.Base,
+  ) {
+    this.workspace = workspace;
+    this.days = days;
+    this.dateToFilter = DateUtilities.getDate(this.days).toISOString();
+    this.client = BitbucketApiClient.createHttpClient({
+      baseUri,
+      username: username,
+      password: password,
+      apiClientName: "Bitbucket API",
+      dateToFilter: this.dateToFilter,
+    });
+  }
+
+  private static createHttpClient({
+    baseUri,
+    username,
+    password,
+    apiClientName,
+    dateToFilter,
+  }: IHttpClientParameters) {
+    const client = axios.create({
+      baseURL: baseUri,
+      auth: {
+        username: username,
+        password: password,
+      },
+    });
+
+    client.interceptors.response.use(
+      async (response) => {
+        soosLogger.verboseDebug(apiClientName, `Response Body: ${JSON.stringify(response.data)}`);
+        if (response.data.next) {
+          return await this.handlePagination(response, client, dateToFilter);
+        }
+        return response;
+      },
+      async (error) => {
+        const { config, response } = error;
+        const maxRetries = 3;
+        config.retryCount = config.retryCount || 0;
+
+        if (response?.status === 429 && config.retryCount < maxRetries) {
```

Comment:
is there a specific limit we need to be concerned with?

Reply:
no, documentation is kind of shallow and i've tried myself to get blocked and seems that after 1 minute you're able to query again, there's not a field or header that indicates timeout time. (Even tho it says 1k if you just wait for one minute or even less it let's you continue)

https://support.atlassian.com/bitbucket-cloud/docs/api-request-limits/